Создание всех классов

MikruykovaEkaterina opened the pull request.

MikruykovaEkaterina committed:
Создание всех классов

Dezodemius commented on Company/Company/Employee.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Company
+{
+  internal abstract class Employee
```
Добавить документацию для класса и его членов

Dezodemius commented on Company/Company/Employee.cs:
```diff
@@ -0,0 +1,16 @@
+{
+  internal abstract class Employee
+  {
+    public abstract int ID { get; set; }
```
Разделить члены класса пустыми строками

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace Company
+{
+  internal class EmployeeManager<T> : IEmployeeManager<T> where T : Employee
```
Добавить документацию для класса и его членов

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+  {
+    private List<T> employees = new List<T>();
+
+    public void Add(T employee)
```
Методы, которые реализуют интерфейс нужно вынести в регион "IEmployeeManager"

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+      T? results = employees.SingleOrDefault(emp => emp.ID == employee.ID);
+      if (results != null)
+        throw new EmployeeAdditionException($"Сотрудник с таким id уже добавлен");
+      //все проверки пройдены, можно добавлять
```
Нет нужны в этом комментарии

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+        throw new EmployeeAdditionException("Нельзя добавить пустого сотрудника");
+      T? results = employees.SingleOrDefault(emp => emp.ID == employee.ID);
+      if (results != null)
+        throw new EmployeeAdditionException($"Сотрудник с таким id уже добавлен");
```
Нужно написать с каким именно ID уже добавлен. Так меньше времени потратиться на выяснение причин проблем

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+    public T Get(string name)
+    {
+      T? result = employees.FirstOrDefault(emp => emp.Name == name); //имя не уникально
+      //я бы лучше возвращала список
```
А ты делай get по ID. Тогда не будет нужды в списке, потому что id всегда уникален

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+      T? result = employees.FirstOrDefault(emp => emp.Name == name); //имя не уникально
+      //я бы лучше возвращала список
+      if (result == null) 
+        throw new EmployeeNotFoundException("Сотрудник с таким именем не содержится в списке");
```
Нужно указать с каким именем

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+
+    public void Update(T employee)
+    {
+      T employeeToUpdateInfo = GetById (employee.ID);
```
Убрать пробел после `GetById`

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+    {
+      Console.WriteLine(employee);
+      if (!employees.Contains(employee)) 
+        throw new EmployeeNotFoundException("Такой сотрудник не содержится в списке");
```
Нужно больше информации добавить о сотруднике. Имя и ID будет достаточно

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+
+    public T GetById(int id)
+    {
+      T? result = employees.SingleOrDefault(emp => emp.ID == id); //id должно быть уникально
```
Все комментарии `//` нужно убрать

Dezodemius commented on Company/Company/EmployeeManager.cs:
```diff
@@ -0,0 +1,62 @@
+    {
+      T? result = employees.SingleOrDefault(emp => emp.ID == id); //id должно быть уникально
+      if (result == null) 
+        throw new EmployeeNotFoundException("Сотрудник с таким id не содержится в списке");
```
Нужно указать значение id

Dezodemius commented on Company/Company/FullTimeEmployee.cs:
```diff
@@ -0,0 +1,64 @@
+  internal class FullTimeEmployee: Employee
+  {
+    private int id;
+    public override int ID
```
Каждый член класса должен быть отделён друг от друга пустой строкой

Dezodemius replied:
Свойства и методы, которые реализуют абстрактные члены предка должны быть в регионе "Базовый класс"

Dezodemius commented on Company/Company/FullTimeEmployee.cs:
```diff
@@ -0,0 +1,64 @@
+      set
+      {
+        if (value <= 0)
+          throw new ArgumentException($"Введено недопустимое значение для (ID). ID должен быть положительным");
```
Тоже нужно указать ID, который я ввёл

Dezodemius commented on Company/Company/FullTimeEmployee.cs:
```diff
@@ -0,0 +1,64 @@
+    private int id;
+    public override int ID
+    {
+      get { return id; }
```
Коль set развёрнут, то и get нужно развернуть

Dezodemius commented on Company/Company/FullTimeEmployee.cs:
```diff
@@ -0,0 +1,64 @@
+      return BaseSalary;
+    }
+
+    public FullTimeEmployee(int id, string name, decimal baseSalary)
```
Переместить конструктор в низ класса

Dezodemius commented on Company/Company/FullTimeEmployee.cs:
```diff
@@ -0,0 +1,64 @@
+      BaseSalary = baseSalary;
+    }
+
+    public override string ToString()
```
Вынести в регион "Базовый класс"

Dezodemius commented on Company/Company/IEmployeeManager.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Company
+{
+  internal interface IEmployeeManager<T>
```
Добавить документацию для интерфейса и членов

Dezodemius commented on Company/Company/IEmployeeManager.cs:
```diff
@@ -0,0 +1,16 @@
+{
+  internal interface IEmployeeManager<T>
+  {
+    void Add(T employee);
```
Разделить члены пустыми строками

Dezodemius commented on Company/Company/InputHelper.cs:
```diff
@@ -0,0 +1,95 @@
+
+namespace Company
+{
+  internal class InputHelper
```
Для класса и методов добавить документацию

Dezodemius commented on Company/Company/MyExceptions.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace Company
+{
+  public class EmployeeCreationException : Exception
```
Вынести исключения в отдельные классы

Dezodemius replied:
Добавить документацию для класса и членов

Dezodemius commented on Company/Company/PartTimeEmployee.cs:
```diff
@@ -0,0 +1,83 @@
+
+namespace Company
+{
+  internal class PartTimeEmployee : Employee
```
Добавить документацию, разделители между членами класса, и регионы

Dezodemius commented on Company/Company/Program.cs:
```diff
@@ -0,0 +1,266 @@
+
+namespace Practice4
+{
+  internal class Program
```
Документация

Dezodemius reviewed (changes requested):
В основном замечания по соглашению о кодировании

Dezodemius commented on Company/Company/InputHelper.cs:
```diff
@@ -0,0 +1,95 @@
+      }
+      catch (ArgumentException ex)
+      {
+        throw new EmployeeCreationException(ex.Message);
```
Плохая практика так делать. Ты теряешь стек вызовов прошлого исключения. Предлагаю просто делать catch исключения и пробрасывать его дальше через throw. 

```
catch (<какой-то тип исключения>)
{
  throw;
}

```

Можно что-то в консоль написать (хотя этот варик не оч)

Dezodemius replied:
И так везде по коду

Dezodemius commented on Company/Company/InputHelper.cs:
```diff
@@ -0,0 +1,95 @@
+    {
+      try
+      {
+        int id = _id == -1 ? ReadInt("ID"): _id;
```
Перед ":" добавить пробел

Dezodemius commented on Company/Company/InputHelper.cs:
```diff
@@ -0,0 +1,95 @@
+      Console.Write($"{prompt}: ");
+      string? value = Console.ReadLine();
+      if (string.IsNullOrWhiteSpace(value))
+        throw new ArgumentException($"Введено недопустимое значение для ({prompt})");
```
а кто вообще может выкинуть ArgumentException?

Dezodemius commented on Company/Company/InputHelper.cs:
```diff
@@ -0,0 +1,95 @@
+  internal class InputHelper
+  {
+
+    public static string ReadString(string prompt)
```
Кажется, это хороший пример для применения дженериков, чтобы не плодить однообразные методы

MikruykovaEkaterina committed:
Добавила документацию

Также добавила абстрактный метод ToString(), чтобы наследники точно переопределяли

MikruykovaEkaterina committed:
Вынесла исключение в отдельный класс

Вынесла исключение EmployeeAdditionException в отдельный класс. Добавила документацию

MikruykovaEkaterina committed:
Вынесла исключение в отдельный класс

Вынесла исключение EmployeeCreationException в отдельный класс. Добавила документацию

MikruykovaEkaterina committed:
Вынесла исключение в отдельный класс

Вынесла исключение EmployeeCreationException в отдельный класс. Добавила документацию

MikruykovaEkaterina committed:
Добавила документацию

Добавила документацию, регионы. сделала исключения более информативными

MikruykovaEkaterina committed:
Добавила документация

Добавила документацию, регионы. Сделала исключения более информативными

MikruykovaEkaterina committed:
Добавила документацию

MikruykovaEkaterina committed:
Добавила документацию

Добавила документацию, сделала исключения более информативными

MikruykovaEkaterina committed:
Update PartTimeEmployee.cs

Добавила документацию, регионы. Сделала исключения более информативными

MikruykovaEkaterina committed:
Добавила документацию

Добавила документацию, убрала комментарии, сделала поиск только по id

MikruykovaEkaterina committed:
Создала дженерик

MikruykovaEkaterina committed:
Теперь пробрасываю исключения

MikruykovaEkaterina committed:
Update Program.cs
